python/catkin/builder.py

```diff
@@ -286,10 +287,42 @@ def get_multiarch():
     return out.strip()
 
 
-def get_python_install_dir():
-    # this function returns the same value as the CMake variable PYTHON_INSTALL_DIR from catkin/cmake/python.cmake
-    relevant_subfolders = 2 if os.name == 'nt' else 3
-    return os.sep.join(get_python_lib().split(os.sep)[-relevant_subfolders:])
+def get_python_install_dir_tuple(cmake_args=None):
+    # this function invokes the CMake file catkin/cmake/python.cmake to get
+    # the same PYTHON_INSTALL_DIR that a package would
+    cmake_path = get_cmake_path()
+    python_cmake_path = os.path.abspath(os.path.join(cmake_path, 'python.cmake'))
+    cmd = ' '.join(['cmake', '-P', python_cmake_path] + (cmake_args or []))
+    working_dir = tempfile.mkdtemp()
+    try:
+        out = subprocess.check_output(cmd, shell=True).decode()
+        m = re.search('.*PYTHON_EXECUTABLE: (.+)$', out, re.MULTILINE)
+        if not m:
+            raise RuntimeError("Failed to get the PYTHON_EXECUTABLE from python.cmake")
+        python_executable = m.group(1)
+        m = re.search('.*PYTHON_INSTALL_DIR: (.+)$', out, re.MULTILINE)
+        if not m:
+            raise RuntimeError("Failed to get the PYTHON_INSTALL_DIR from python.cmake")
+        return (python_executable, m.group(1))
+    finally:
+        shutil.rmtree(working_dir)
+
+__cached_python_install_dir_tuple = {}
+
+
+def get_python_install_dir_tuple_cached(cmake_args=None):
+    global __cached_python_install_dir_tuple
+    if cmake_args not in __cached_python_install_dir_tuple:
+        __cached_python_install_dir_tuple[cmake_args] = get_python_install_dir_tuple(cmake_args)
+    return __cached_python_install_dir_tuple[cmake_args]
+
+
+def get_python_install_dir(cmake_args=None):
+    return get_python_install_dir_tuple(cmake_args)[1]
+
+
+def get_python_install_dir_cached(cmake_args=None):
+    return get_python_install_dir_tuple_cached(cmake_args)[1]
```

Comment:
I think it is clearer to have a "cached" version of the function rather having a parameter to control the behavior. I left the `get_python_install_dir` function (rather than replace it completely with `get_python_install_dir_tuple`) since someone might be using this function already.